src/main/java/io/cryostat/recordings/RecordingTargetHelper.java

```diff
@@ -314,6 +339,23 @@ public IRecordingDescriptor startRecording(
                 .findFirst();
     }
 
+    public void notifyRecordingStopped(String recordingName, String targetId) {
+        notificationFactory
+                .createBuilder()
+                .metaCategory(STOP_NOTIFICATION_CATEGORY)
+                .metaType(HttpMimeType.JSON)
+                .message(Map.of("recording", recordingName, "target", targetId))
+                .build()
+                .send();
+    }
+
+    public synchronized void cancelScheduledNotificationIfExists(String stoppedRecordingName) {
```

Comment:
Looks better. I would be cautious of using the simple `synchronized` method-level keyword here. This uses the `RecordingTargetHelper` instance itself as the lock instance. For now, since the only two places where that lock instance is used are the two notifications-related methods, that's fine. But if in the future we decide we need to synchronize accesses to some other resource in here - say the `reportService` for some reason - then we don't want those two separate resources causing threads to contend for the same lock when they really are likely to be able to both proceed safely.

I think it might be better ("safer" as well as more explicit) to use the map instance as the lock.

```java
void cancelNotification(String name) {
  synchronized (notifications) {
    var f = notifications.remove(name);
    if (f != null) f.cancel(true);
  }
}

void scheduleNotification(String name) {
  synchronized (notifications) {
    notifications.put(name, scheduler.schedule());
  }
}

void stopRecording(String name) {
  synchronized (notifications) {
    tcm.executeConnectedTask(() -> cancelTheRecording());
    cancelNotification(name);
  }
}

void startRecording(String name) {
  synchronized (notifications) {
    tcm.executeConnectedTask(() -> startRecording());
    scheduleNotification(name);
  }
}
```

Reply:
> But if in the future we decide we need to synchronize accesses to some other resource in here - say the reportService for some reason - then we don't want those two separate resources causing threads to contend for the same lock when they really are likely to be able to both proceed safely.

Thanks, this makes sense.

In your example, why does the call to `scheduleNotification` in `startRecording()` need to be wrapped in a synchronized block as well as inside the `scheduleNotification()` method itself? Is it because `startRecording` or `stopRecording` need to acquire the lock before entering `scheduleNotification`?

Reply:
I was thinking it makes sense that no other thread should be modifying the notifications map while there is a thread actively working on connecting to a target and starting/stopping a recording. If we only lock within `scheduleNotification` then two threads can both be working on connecting to a target with one starting a recording and the other trying to stop the same recording erroneously, but depending on thread scheduling we can end up with the recording starting, then being stopped, then the stop notification being scheduled. If we lock earlier, before even connecting to the target, then this can't happen - the stop can't interleave between the start and the scheduling.

Perhaps this should actually be a map of locks, with a lock for each targetId, so we don't end up reducing request throughput here when making parallel requests to separate targets. Acquiring such a target-specific lock before opening the connection to the target, then connecting, starting/stopping the recording, and emitting/scheduling a stop notification preserves the nice atomicity we want here without also creating a chokepoint where threads are blocked waiting for each other to do things on different targets.

Reply:
It looks like starting/stopping a recording and scheduling a stop notification always calls `executeConnectedTask()`, so I decided to put the map of locks in TargetConnectionManager. `executeConnectedTask()` will aqcuire a target lock before calling `task.execute()`.